nav2_util/src/robot_utils.cpp

```diff
@@ -0,0 +1,104 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <nav2_util/robot_utils.hpp>
+
+namespace nav2_util
+{
+
+VelocityPublisher::VelocityPublisher(rclcpp::Node::SharedPtr & node,
+    const std::string topic) {
+    vel_pub_ = node->create_publisher<geometry_msgs::msg::Twist>(topic, 1);
+  }
+
+void 
+VelocityPublisher::publishCommand(const geometry_msgs::msg::Twist& cmd_vel) {
+    vel_pub_->publish(cmd_vel);
+}
+
+RobotStateHelper::RobotStateHelper(rclcpp::Node::SharedPtr & node,
+  const std::string odom_topic) : node_(node) {
+  pose_sub_ = node_->create_subscription<geometry_msgs::msg::PoseWithCovarianceStamped>(
+    "amcl_pose", std::bind(&RobotStateHelper::onPoseReceived, this, std::placeholders::_1));
+
+  odom_sub_ = node_->create_subscription<nav_msgs::msg::Odometry>(
+    odom_topic, std::bind(&RobotStateHelper::onOdomReceived, this, std::placeholders::_1));
+
+  initial_odom_received_ = false;
+  initial_pose_received_ = false;
+}
+
+bool
+RobotStateHelper::getOdometry(nav_msgs::msg::Odometry::SharedPtr & robot_odom) {
+  std::shared_lock<std::shared_timed_mutex> lock(state_mutex_);
+  if (!initial_odom_received_) {
+    RCLCPP_DEBUG(node_->get_logger(),
+      "getOdometry: Can't return current velocity: Initial odometry not yet received.");
+    return false;
+  }
+
+  robot_odom = current_odom_;
+  return true;
+}
+
+bool
+RobotStateHelper::getCurrentPose(geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr & robot_pose,
+  const bool use_topic) {
+  if (use_topic) {
+    return getGlobalLocalizerPose(robot_pose);
+  }
+
+  return getTfPose(robot_pose);
+}
+
+void
+RobotStateHelper::onPoseReceived(geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr robot_pose) {
+  std::unique_lock<std::shared_timed_mutex> lock(state_mutex_);
+  current_pose_ = robot_pose;
+  if (!initial_pose_received_) {
+    initial_pose_received_ = true;
+  }
+}
+
+void
+RobotStateHelper::onOdomReceived(nav_msgs::msg::Odometry::SharedPtr msg) {
+  std::unique_lock<std::shared_timed_mutex> lock(state_mutex_);
+  current_odom_ = msg;
+  if (!initial_odom_received_) {
+    initial_odom_received_ = true;
+  }
+}
+
+bool
+RobotStateHelper::getGlobalLocalizerPose(geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr & robot_pose) {
+  std::shared_lock<std::shared_timed_mutex> lock(state_mutex_);
+  if (!initial_pose_received_) {
+    RCLCPP_DEBUG(node_->get_logger(),
+      "Robot: Can't return current pose: Initial pose not yet received.");
+    return false;
+  }
+
+  robot_pose = current_pose_;
+  return true;
+}
+
+bool
+RobotStateHelper::getTfPose(geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr & /*robot_pose*/) {
```

Comment:
Were you planning to implement this function as is done [currently in Costmap2DROS](https://github.com/ros-planning/navigation2/blob/178c7c6ce4aa3ea58189302e1287d44f800543cd/nav2_costmap_2d/src/costmap_2d_ros.cpp#L435)? If we were to merge this before implementing, I would at least put the use_topic default as true

Reply:
I didnt want to mix too much stuff into this PR since this PR is primarily a refactor. Adding new functionality I think should be done after merging. Default `true` is right, thanks